### PR TITLE
Adding new `application_gateway` data_source

### PR DESCRIPTION
**Pull request by petems**
Closes https://github.com/terraform-providers/terraform-provider-azurerm/issues/5030

Draft for now, as I want to make sure I'm making the data_source correctly, after that I'll add the rest of the attributes

### CODE REVIEW

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @petems

Thanks for opening this PR.

I've taken a look through and this is off to a good start - I've taken a look through and left some comments inline, as mentioned offline this'll need to be registered here to be usable in the provider: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/network/registration.go#L24

Thanks!

File: azurerm/internal/services/network/registration.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ func (r Registration) WebsiteCategories() []string {
 // SupportedDataSources returns the supported Data Sources supported by this Service
 func (r Registration) SupportedDataSources() map[string]*schema.Resource {
 	return map[string]*schema.Resource{
+		"azurerm_application_gateway":                       dataSourceApplicationGateway(),
```
Comment (petems) [inline azurerm/internal/services/network/registration.go:24]: @tombuildsstuff Is this correct?

File: azurerm/internal/services/network/application_gateway_data_source_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+func (AppGatewayDataSource) complete(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_application_gateway" "test" {
+  resource_group_name = azurerm_application_gateway.test.resource_group_name
+  name                = azurerm_application_gateway.test.name
+}
+`, AppGatewayDataSource{}.complete(data))
```
[the remark this answers is not in the document]
Comment (petems) [reply azurerm/internal/services/network/application_gateway_data_source_test.go]: Done! 👍

File: azurerm/internal/services/network/application_gateway_data_source.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+			return fmt.Errorf("Error: Application Security Group %q (Resource Group %q) was not found", name, resourceGroup)
+		}
+
+		return fmt.Errorf("Error making Read request on Application Security Group %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resourceGroup)
```
[the remark this answers is not in the document]
Comment (petems) [reply azurerm/internal/services/network/application_gateway_data_source.go]: Good point, fixed

**Comment by petems** [comment]
I've pulled most of the logic from the resource, so there's a lot of duplicted code here, is there value in refactoring this into a shared librariy? If so, is there an example out there?

File: azurerm/internal/services/network/registration.go (hunk 1 above)
Comment (katbyte) [reply azurerm/internal/services/network/registration.go:24]: Yep this is!

**Comment by katbyte** [comment]
@petems - we're going to need to flip all the properties to computed as this is a data source

File: azurerm/internal/services/network/application_gateway_data_source.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+			return nil
+		}
+
+		return fmt.Errorf("Error making Read request on Application Gateway %s: %+v", name, err)
+	}
+
+	if location := applicationGateway.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	d.Set("zones", applicationGateway.Zones)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/network/application_gateway_data_source.go:65]: this data source only defines `name, location, resource_group_name, tags` above, these other fields don't exist (so will raise an error) - can we remove them, or add the additional fields to the schema?

File: azurerm/internal/services/network/application_gateway_data_source_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+func (AppGatewayDataSource) basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_application_gateway" "test" {
+  resource_group_name = azurerm_application_gateway.test.resource_group_name
+  name                = azurerm_application_gateway.test.name
+}
+`, AppGatewayResource{}.basic(data))
```
Comment (tombuildsstuff) [inline azurerm/internal/services/network/application_gateway_data_source_test.go:37]: sorry, my bad - the type name for this is

```suggestion
`, ApplicationGatewayResource{}.basic(data))
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @petems 

Thanks for pushing those changes - I've taken a look through and left some comments inline, if we can get those sorted (and subsequently add/generate some documentation, there's details on how to generate that in the readme) - then we should be able to run the tests here 👍 

Thanks!

**Comment by tombuildsstuff** [comment]
Tests look good 👍

**Pull request merged**